### ISSUE TITLE
New DICOM variant (insufficient image data @ error/dcm.c/ReadDCMImage/3309)

### ISSUE THREAD

**Issue by Malvineous** [ImageMagick/ImageMagick#2534]
Just encountered a DICOM medical image that I'm unable to read with ImageMagick.  Here's an anonymised sample to reproduce the problem: [sample.dcm.zip](https://github.com/ImageMagick/ImageMagick/files/5193797/sample.dcm.zip)

    $ identify sample.dcm
    identify: insufficient image data in file `sample.dcm' @ error/dcm.c/ReadDCMImage/3309.
    $ convert sample.dcm out.png
    convert: insufficient image data in file `sample.dcm' @ error/dcm.c/ReadDCMImage/3309.

Would it be possible to add support for these images to the DICOM reader?

GIMP will open the files but it looks like it's reading the 16-bit data as 8-bit or 1-bit, as the image comes out like it has had a noise filter applied over it then been converted to monochrome.  The sample image is of a chest X-ray.

**Comment by urban-warrior** [comment]
Thanks for alerting us to this problem.  It looks like we need to update the DICOM data dictionary.  We do not have an ETA on a patch.

**Comment by Malvineous** [comment]
Thanks for the update.  I've delved into the code and it looks like the problem is that the file has multiple images.  The first one has dimensions of 64x64 which seems to read ok, but when it gets to the main image, it still has these now incorrect dimensions.  When it reads the pixel data, the block is 8 MB in size, but the code only reads 4 kB of data (64x64 16-bit pixels).  It does not then discard the unused pixel data, but rather resumes reading the next DICOM header right in the middle of the pixel data.

Although the real problem is that it is not picking up the correct image dimensions, I think it should at least skip to the end of the pixel data just in case it ever encounters a file with trailing data or padding following the pixel data.  The error message above is a result of the code getting confused and trying to read the pixel data as DICOM headers.

**Comment by Malvineous** [comment]
Ok, so adding a `DiscardBlobBytes()` call to the end of the pixel data processing allowed it to skip over the unused data in the DICOM block and stopped it giving me an error.

I have now spotted the problem.  The DICOM attributes are ordered like this:

1. dimensions = 2101x2101
2. dimensions = 64x64
3. pixel data for 64x64 image
4. pixel data for 2101x2101 image

GIMP works because it only uses the first set of dimensions, ImageMagick fails because it uses the last set of dimensions.

There must be some way in the DICOM format to match up the dimensions against the correct pixel data block, but I'm not sure exactly how.  But that at least appears to be the underlying problem.

**Comment by Malvineous** [comment]
I figured out the problem and did a PR for it.  Hopefully it's an acceptable way of dealing with the issue.

Turns out there are sequences in DICOM files that work a bit like folders, in that attributes (like image dimensions) that are changed inside a sequence/folder aren't supposed to persist past the end of the sequence, when processing returns to the parent "folder".

I used a linked list as a stack to save and restore the current values as the sequences begin and end, and this at least allowed my image above to be read correctly.
